### PR TITLE
Move constraint editor label HTML out of the database

### PR DESCRIPTION
**Pull request by jennifer-richards**
Icons to indicate constraint violations between sessions are currently rendered based on HTML in the `ConstraintName` table of the DB. At least one of these involves interpolating a violation count into the rendered string. Further, at run-time a couple additional non-persistent `ConstraintName` instances are created.

This puts HTML that must be coordinated with HTML templates, imported icon libraries, and JS code in the database and in Python utility code. Additionally, logic involved in structuring the data relies on rendering the icons into strings and using these as dictionary keys.

This pull request moves the HTML out of the database and utility code and into a template tag, `constraint_icon_for`, which takes a `ConstraintName` and optional `count` parameter and returns the appropriately rendered code. Use of the rendered HTML in the dictionary keys is replaced with a `NamedTuple` containing these values.

There is additional work to clean this up architecturally, but this is intended to make things slightly better while we sort out when/how to address the bigger questions.

### CODE REVIEW

File: ietf/meeting/templatetags/editor_tags.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# Copyright The IETF Trust 2022, All Rights Reserved
+# -*- coding: utf-8 -*-
+
+"""Custom tags for the schedule editor"""
+import debug  # pyflakes: ignore
```
Comment (meadmaker) [inline ietf/meeting/templatetags/editor_tags.py:5]: Should this be removed before going into the mainline code?

File: ietf/meeting/templatetags/tests.py [hunk 2]
```diff
@@ -1,10 +1,13 @@
 # Copyright The IETF Trust 2009-2020, All Rights Reserved
 # -*- coding: utf-8 -*-
+import debug  # pyflakes: ignore
```
Comment (meadmaker) [inline ietf/meeting/templatetags/tests.py:3]: Should this be removed before going into the mainline code?

File: ietf/meeting/templatetags/tests.py [hunk 3]
```diff
@@ -43,3 +46,48 @@ def test_location_anchor_node(self):
...
+class EditorTagsTests(TestCase):
+    def _supported_constraint_names(self):
+        """Get all ConstraintNames that must be supported by the tags"""
+        constraint_names = set(ConstraintName.objects.filter(used=True))
+        # Instantiate a couple that are added at run-time in meeting.utils
+        constraint_names.add(ConstraintName(slug='joint_with_groups', name='joint with groups'))
+        constraint_names.add(ConstraintName(slug='responsible_ad', name='AD'))
+        # Reversed names are also added at run-time
+        reversed = [
+            ConstraintName(slug=n.slug + "-reversed", name="{} - reversed".format(n.name))
```
Comment (meadmaker) [inline ietf/meeting/templatetags/tests.py]: Why does the slug get string concatenation while the name gets string interpolation?

File: ietf/meeting/templatetags/tests.py [hunk 4]
```diff
@@ -43,3 +46,48 @@ def test_location_anchor_node(self):
...
+        result = template.render(Context({cn.slug.replace('-', '_'): cn for cn in constraint_names}))
+        q = PyQuery(result)
+        for cn in constraint_names:
+            elts = q(f'div#{cn.slug}')
+            self.assertEqual(len(elts), 1)
+            self.assertGreater(len(elts.html()), 0)
+
+            elts = q(f'div#{cn.slug}-count')
+            self.assertEqual(len(elts), 1)
+            self.assertGreater(len(elts.html()), 0)
```
Comment (meadmaker) [inline ietf/meeting/templatetags/tests.py]: This seems like a pretty minimal test - just that the `<div>`'s wind up with anything in them.  Does it deserve some sort of test for, say, whether it rendered an icon, or whether it rendered the correct icon, or such?

File: ietf/meeting/views.py [hunk 5]
```diff
@@ -558,20 +558,20 @@ def prepare_sessions_for_display(sessions):
...
-                        constrained_sessions_grouped_by_label[format_html(constraint_name.formatted_editor_label, count=count)].add(session_pk)
-
-                else:
-                    constrained_sessions_grouped_by_label[constraint_name.formatted_editor_label].update(session_pks)
-
-            s.constrained_sessions = list(constrained_sessions_grouped_by_label.items())
+                for session_pk, grouped_session_pks in itertools.groupby(session_pks):
+                    # count is the number of instances of session_pk - should only have multiple in the
+                    # case of bethere constraints, where there will be one per person.pk
+                    count = sum(1 for i in grouped_session_pks)
```
Comment (meadmaker) [inline ietf/meeting/views.py]: Is there really nothing like `length(grouped_session_pks)` or `grouped_session_pks.size()` or anything?

**Review by meadmaker — APPROVED** [review]
Looks like it should work.  I have some questions, but if you want to submit this as-is anyway, go ahead.

File: ietf/meeting/templatetags/tests.py (hunk 3 above)
Comment (jennifer-richards) [reply ietf/meeting/templatetags/tests.py]: Good question but I can't answer it. It was copy/pasted from `utils.py`.

File: ietf/meeting/templatetags/editor_tags.py (hunk 1 above)
Comment (jennifer-richards) [reply ietf/meeting/templatetags/editor_tags.py:5]: The convention in the project is to leave them with the `pyflakes: ignore` comment. The `debug` module is nopped out when run in production.

File: ietf/meeting/templatetags/tests.py (hunk 2 above)
Comment (jennifer-richards) [reply ietf/meeting/templatetags/tests.py:3]: See above.

File: ietf/meeting/templatetags/tests.py (hunk 4 above)
Comment (jennifer-richards) [reply ietf/meeting/templatetags/tests.py]: I went back and forth on that. The aim here is to make sure that there aren't any known constraint names that are not handled at all. The icons are somewhat varied in structure - some are `<i>` tags and others are styled text, so the test winds up being very nearly a duplicate of the module being tested.

Probably worth adding in the redundancy, though, so I'll do that.

File: ietf/meeting/views.py (hunk 5 above)
Comment (jennifer-richards) [reply ietf/meeting/views.py]: There's `len(list(grouped_session_pks))` which is probably a bit clearer and somewhat idiomatic. The issue is that `itertools.groupby()` operates on iterators which are not required to have finite length.

File: ietf/meeting/templatetags/tests.py (hunk 3 above)
Comment (jennifer-richards) [reply ietf/meeting/templatetags/tests.py]: For equity's sake, I'm replacing both with f-string literals. They were both wrong.

**Pull request merged**